src/cast.rs

```diff
@@ -154,3 +155,36 @@ where
     /// won't need to call this.
     fn unchecked_from_js_ref(val: &JsValue) -> &Self;
 }
+
+/// Trait implemented for wrappers around `JsValue`s generated by `#[wasm_bindgen]`.
+#[doc(hidden)]
+pub trait JsObject: JsCast + JsUpcastRef + WasmDescribe {}
+
+/// A trait for upcasting generic reference types to `&JsValue`.
+///
+/// For example, an imported function that accepts a generic `&T` argument
+/// may use `JsUpcastRef` to convert `&T` into `&JsValue`.
+///
+/// Unlike `JsCast`, which requires types to already be able to be represented
+/// as a JsValue, `JsUpcastRef` is able to apply to arbitrary primitive types
+/// that otherwise would require an explicit conversion. This is done by providing
+/// constant reference slots for JS values available only at the generic bindgen
+/// call interface, by referencing constant slots in __rt.
+///
+/// By setting the `USES_VALUE_SLOTS` constant, the bindgen is then able to inform
+/// the bindgen process that it is using these externref slots and that they should
+/// be cleared after the function call, guaranteeing the JsValue indices are
+/// live for as long as the function call only.
+///
+/// This trait is automatically implemented for any type supporting `AsRef<JsValue>`.
+pub trait JsUpcastRef<T = JsValue> {
```

Comment:
I remember you initially added this machinery for references when you thought we can't accept both references and plain values in the same signature.

Now that we figured it out, can we leave this trait and slots implementation out of this PR? It feels like a large chunk of complexity here comes just from this reference upcasting, even though I imagine in real world scenarios it's going to be relatively uncommon for someone to want to pass `&i32` instead of `i32` and so on, and removing it from this PR would reduce scope to just generics for types we support today.

And if it's not uncommon, we can always add it separately, as being able to pass references to more types is not strictly tied to generics support.

Reply:
While that is possible, the problem is the requirement for type specialization on the generic - where a singular specialization is possible on `&T` as `$JsValue` requiring no further type hints making it backwards compatible.

Reply:
In case that wasn't clear - if `T` can be either `&JsValue` or `JsValue` then more type information is required from the usage context, and thus this requirement on the type is a breaking change for existing usage.

Reply:
I don't understand, can you show an example that would work today but break with such generic?

Reply:
Actually I think the problem is simpler:

```
extern "C" { 
    #[wasm_bindgen(static_method_of = Promise)]
    pub fn resolve(obj: &JsValue) -> Promise;
}
```

Can only be represented by the generic:

```
extern "C" { 
    #[wasm_bindgen(static_method_of = Promise)]
    pub fn resolve<T>(obj: &T) -> Promise<T>;
}
```

It cannot be represented by the generic:

```
extern "C" { 
    #[wasm_bindgen(static_method_of = Promise)]
    pub fn resolve<T>(obj: T) -> Promise<T>;
}
```

as when we generate the specialization by substituting JsValue that would be a breaking change.

Reply:
Ah wait I understand what you mean. Hmm. Need to think about this.

I'd really like to avoid having to use references for primitives. Maybe there is some solution with custom trait that would have T::Output that would unwrap reference automatically...

Let's discuss more next week.

Reply:
This doesn't create a requirement for using references for primitives - most APIs should use value types for primitives, and also support `&T` or `T` interchangeably. But we need to special case `&T` for this backwards compatibility.

Ensuring `T` supports a ref case could be something we could add for new APIs that want to support own or ref, but that would require a separate design. If that's the case you're interested in happy to discuss it further.

Reply:
`JsUpcastRef` has now been removed and replaced with `&JsValue<T>`.